fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/synthetic/SyntheticRecordFromStoredRecordPlan.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * SyntheticRecordFromStoredRecordPlan.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.synthetic;
+
+import com.apple.foundationdb.API;
+import com.apple.foundationdb.record.ExecuteProperties;
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore;
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoredRecord;
+import com.apple.foundationdb.record.provider.foundationdb.FDBSyntheticRecord;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Set;
+
+/**
+ * A plan for deriving synthetic records from a given record.
+ *
+ */
+@API(API.Status.EXPERIMENTAL)
+public interface SyntheticRecordFromStoredRecordPlan extends PlanHashable  {
```

Comment:
Yes, the goal is that the inside of `SyntheticRecordPlanner` should be able to use the new planner. If the new planners representation of join results is cursors of bindings or of compound record objects, then that might make sense. If those are formed at a higher level, then this could continue to be separate.

As for the specifics of passing in the start record, I think that would be as a specially named binding if things did get unified.